src/vm/tieredcompilation.cpp

```diff
@@ -98,71 +96,81 @@ NativeCodeVersion::OptimizationTier TieredCompilationManager::GetInitialOptimiza
     WRAPPER_NO_CONTRACT;
     _ASSERTE(pMethodDesc != NULL);
 
-#ifdef FEATURE_TIERED_COMPILATION
+#if defined(FEATURE_TIERED_COMPILATION) && !defined(DACCESS_COMPILE)
+    if (!pMethodDesc->IsEligibleForTieredCompilation())
+    {
+        // The optimization tier is not used
+        return NativeCodeVersion::OptimizationTier0;
+    }
+
     if (pMethodDesc->RequestedAggressiveOptimization())
     {
         // Methods flagged with MethodImplOptions.AggressiveOptimization begin at tier 1, as a workaround to cold methods with
         // hot loops performing poorly (https://github.com/dotnet/coreclr/issues/19751)
         return NativeCodeVersion::OptimizationTier1;
     }
-#endif // FEATURE_TIERED_COMPILATION
+
+    if (!g_pConfig->TieredCompilation_CallCounting())
+    {
+        // Call counting is disabled altogether through config, the intention is to remain at the initial tier
+        return NativeCodeVersion::OptimizationTier0;
+    }
+
+    if (!pMethodDesc->GetCallCounter()->IsTier0CallCountingEnabled(pMethodDesc))
```

Comment:
I like the idea of having an `Unknown` value and making a single transition from `Unknown` to `0` or `1`, but there doesn't seem to be a good final place at the moment where the transition can be made and the value locked for future changes. Suppose while one thread (T0) is in the process of processing the first call to the method and another thread (T1) wants to try to transition from `Unknown` to `1` at the same time. If T0 chooses to finalize the transition from `Unknown` to `0` after jitting it, it is still not safe for `T1` to see that the value is `Unknown` and transition to `1` because `T0` may have already finished generating tier 0 code but not finalized the transition from `Unknown` to `0`. In order to do that before jitting, the value would have to transition to a `PendingDetermination` or something like that before jitting (to prevent changes from other threads) and then to `0` after jitting. At the moment there aren't any cases where other threads would try to transition to `1` so it's not an issue. However, making the change from `Unknown` to `0` would involve extra work in a hot path during startup, which is unnecessary at the moment and we may as well leave it at `Unknown` which would imply `0` anyway and it would be effectively the same as how it is now. I think it's ok for a threads that are about to JIT the method for the first time to allow changes from `0` to `1` in a deterministic fashion such that races between those threads would not be an issue. A different thread may not do so as freely, and having call counting occur prior to jitting makes some information available (the existence of the call counting entry) available to prevent the other thread from doing so.

I have a side branch that refactors all of the call counting stuff into `PublishVersionableCodeIfNecessary`, unfortunately even though my changes should actually improve startup time if anything, I was seeing a regression in startup time in places that had nothing to do with my changes. I left further investigation of that for later.

It could use some more thinking about, I'd prefer to leave this change for later.

Reply:
I'm OK if you leave this as-is for now. Also it would be good to think about what happens if someone uses SOS to view the code version before it is jitted and is incorrectly lead to believe that its a tier0 version when really it is an indeterminate version at that point.

> Suppose while one thread (T0) is in the process of processing the first call to the method and another thread (T1) wants to try to transition from `Unknown` to `1` at the same time.

I would eliminate this possibility by ensuring its a lazy computation, but it only has one correct answer regardless of which thread does the computation. I believe all policies we've looked at so far have that policy.

> However, making the change from `Unknown` to `0` would involve extra work in a hot path during startup.

Not necessarily, but agree a simple implementation would. Given how high overhead our current call counter is, I'm skeptical that setting a bit in the call-counter has measurable overhead but for sake of argument I'll assume I'm wrong on that. If we wanted to make a more optimized implementation, right now there is already a natural mapping that if the call-counter dictionary entry doesn't exist that is an unknown state, if it exists enabled that is tier0, and if it exists disabled that is tier1. So our tier0 code goes unknown -> 0, and our tier1 code goes unknown -> 0 -> 1. By delaying the creation of the entry until after the code was jitted the tier1 case could be optimized to initialize directly into the disabled state and the tier0 case remains the same other than happening a little later. Of course if we really wanted this to be faster we'd switch the call counting scheme entirely, but I still think an unknown state won't saddle us perf trouble.

Reply:
> Also it would be good to think about what happens if someone uses SOS to view the code version before it is jitted and is incorrectly lead to believe that its a tier0 version when really it is an indeterminate version at that point.

At the moment, the final determination of the tier is made before the JIT completes so if it sees a native code pointer it will also get the correct tier. After adding `Unknown` an appropriate point to finalize the tier may be before the native code is set.

> I would eliminate this possibility by ensuring its a lazy computation, but it only has one correct answer regardless of which thread does the computation. I believe all policies we've looked at so far have that property.

The information doesn't have to be available from the MethodDesc, it could be based on how it's used, for instance it is reasonable to consider that if disabling tier 0 JIT for any method that has loops does not pan out, that tier 0 JIT may be disabled for the start method of a thread. Anyway all I meant is there should be a good spot to make the final transition and that will need more thinking.

> Given how high overhead our current call counter is, I'm skeptical that setting a bit in the call-counter has measurable overhead but for sake of argument I'll assume I'm wrong on that.

In a comparison between no tiering and various modes of tiering, they both have a high overhead in the prestub, so a small addition relative to the prestub is not necessarily small relative to the diff of the comparison. I agree with the likelihood that it won't be significant, but it's not worth going through the trouble until the refactor.

Reply:
There is already an easily noticeable startup overhead from going through `PublishableVersionableCodeIfNecessary`. It is typically more than compensated for by tier 0 jitting, so not a big deal overall.